client/src/send_and_confirm_transactions_in_parallel.rs

```diff
@@ -0,0 +1,465 @@
+                match e.kind {
+                    solana_rpc_client_api::client_error::ErrorKind::Io(_)
+                    | solana_rpc_client_api::client_error::ErrorKind::Reqwest(_) => {
+                        // continue on io error, we will retry the transaction
+                    }
+                    solana_rpc_client_api::client_error::ErrorKind::TransactionError(e) => {
+                        error_map.insert(*index, e);
+                        continue;
+                    }
+                    _ => {
+                        return Err(e.into());
+                    }
+                }
+            }
+        }
+        // send to confirm the transaction
+        transaction_map.insert(
+            signature,
+            TransactionData {
+                index: *index,
+                serialized_transaction,
+                last_valid_blockheight: blockhashdata.last_valid_blockheight,
+                message: message.clone(),
+            },
+        );
+
+        if let Some(progress_bar) = progress_bar {
+            set_message_for_confirmed_transactions(
+                progress_bar,
+                confirmed_transactions.load(std::sync::atomic::Ordering::Relaxed),
+                total_transactions,
+                None,
+                blockhashdata.last_valid_blockheight,
+                &format!(
+                    "Sending {}/{} transactions",
+                    counter + 1,
+                    current_transaction_count,
+                ),
+            );
+        }
+    }
+    Ok(())
+}
+
+async fn confirm_transactions_till_block_height_and_resend_unexpired_transaction_over_tpu(
+    progress_bar: &Option<indicatif::ProgressBar>,
+    tpu_client: &Option<TpuClient<QuicPool, QuicConnectionManager, QuicConfig>>,
+    transaction_map: Arc<DashMap<Signature, TransactionData>>,
+    current_blockheight: Arc<AtomicU64>,
+    confirmed_transactions: Arc<AtomicU32>,
+    total_transactions: usize,
+) {
+    let transactions_to_confirm = transaction_map.len();
+    let max_valid_block_height = transaction_map
+        .iter()
+        .map(|x| x.last_valid_blockheight)
+        .max();
+
+    if let Some(mut max_valid_block_height) = max_valid_block_height {
+        if let Some(progress_bar) = progress_bar {
+            set_message_for_confirmed_transactions(
+                progress_bar,
+                confirmed_transactions.load(std::sync::atomic::Ordering::Relaxed),
+                total_transactions,
+                Some(current_blockheight.load(Ordering::Relaxed)),
+                max_valid_block_height,
+                &format!(
+                    "Waiting for next block, {transactions_to_confirm} transactions pending..."
+                ),
+            );
+        }
+
+        // wait till all transactions are confirmed or we have surpassed max processing age for the last sent transaction
+        while !transaction_map.is_empty()
```

Comment:
i think we need to poll the exit signal in this loop?

Reply:
oh can async `abort()` actually preempt a looping task?  i thought the task had to voluntarily yield back to the executor :thinking: